### ISSUE TITLE
Should we get rid of the "optional" attribute from @Inject?

### ISSUE THREAD

**Issue by cgruber** [javax-inject/javax-inject#1]
_From [crazybob...@gmail.com](https://code.google.com/u/114815478402472606661/) on June 16, 2009 13:03:32_

I don't think this feature carries its weight. It's very tricky to specify
and implement. I think it's better to just bind a default value explicitly.

_Original issue: http://code.google.com/p/atinject/issues/detail?id=1_

**Comment by cgruber** [comment]
_From [crazybob...@gmail.com](https://code.google.com/u/114815478402472606661/) on June 16, 2009 13:09:39_

We can always add this feature later.

**Comment by cgruber** [comment]
_From [gavin.k...@gmail.com](https://code.google.com/u/105743409453530897815/) on June 16, 2009 13:23:25_

Please get rid of it. Since we already have the Provider interface it's just not needed.

**Comment by cgruber** [comment]
_From [crazybob...@gmail.com](https://code.google.com/u/114815478402472606661/) on June 16, 2009 15:56:50_

I may have misunderstood your point, but the Provider interface can't supplant
"optional", at least not in Guice.

Guice only skips injections with "optional=true" if the dependency or something it
depends on isn't configured at all. If Guice encounters an error while resolving the
dependency (like an exception thrown from a factory), Guice generates an error
instead of skipping the injection. This ensures that we never ignore real errors.

Further, if you inject Provider<Foo> using Guice and you don't have a binding to Foo
configured, Guice will generate an error immediately (initialization time at the latest).

I'm more than happy to drop "optional" though and will take it out of the spec
shortly unless someone objects.

**Comment by cgruber** [comment]
_From [gavin.k...@gmail.com](https://code.google.com/u/105743409453530897815/) on June 16, 2009 16:01:36_

Oh, OK, I assumed that you would not get the exception until you called get().
Whatever, I still think we don't need it :-)

**Comment by cgruber** [comment]
_From [thiag...@gmail.com](https://code.google.com/u/103689518864398263449/) on June 16, 2009 17:26:05_

I agree that this feature is worth including in the specification.

**Comment by cgruber** [comment]
_From [tpeie...@gmail.com](https://code.google.com/u/116658917155462694901/) on June 16, 2009 19:14:36_

I found it very handy to be able to use optional injection in the implementation of the 
DWR-Guice integration package, and I don't see an easy alternative. So I'm tentatively 
objecting.

**Comment by cgruber** [comment]
_From [limpbizkit](https://code.google.com/u/limpbizkit/) on June 17, 2009 01:31:17_

Optional is handy, especially when you want to supply a module for somebody else to consume...

   @Inject(optional=true) @Port int port = 8080;

**Comment by cgruber** [comment]
_From [gavin.k...@gmail.com](https://code.google.com/u/105743409453530897815/) on June 17, 2009 09:30:20_

I would *much* prefer to handle this usecase by injecting a Provider.

**Comment by cgruber** [comment]
_From [thiag...@gmail.com](https://code.google.com/u/103689518864398263449/) on June 17, 2009 11:46:28_

I'm sorry, I wrote exactly the opposite of what I think: I think this feature is 
*not* worth to be included in the specification.

**Comment by cgruber** [comment]
_From [roberto....@sun.com](https://code.google.com/u/101742655369438269925/) on June 17, 2009 11:59:40_

How is injecting a Provider equivalent?

The idea of @Inject(optional=true) is that the bean will take advantage of a dependency if there is one, 
otherwise it'll do its job using some meaningful default.

If you injected a Provider instead, the bean would always call get(), then it'd have to handle getting a null 
back, or worse an exception. Note that the Provider would look like any other one, so the contract for 
Provider.get() would have to allow not returning an instance. But that's bad: I wouldn't want all beans that use 
Provider to have to defend against nulls or exceptions as a regular occurrence! I guess I'm agreeing with 
Guice's behavior here.

We could keep the feature and still drop optional from @Inject by adding a new type for "faulty providers".

**Comment by cgruber** [comment]
_From [limpbizkit](https://code.google.com/u/limpbizkit/) on June 17, 2009 12:05:35_

Yeah, Provider is not at all an alternative to optional=true. In fact, you can use them together:

  /**
   * If the user has configured an executor service, use that. Otherwise, we
   * fall back to using our own single-threaded executor.
   */
  @Inject(optional=true) Provider<ExecutorService> executorServiceProvider = new 
Provider<ExecutorService>() {
    ExecutorService result = null;
    public synchronized ExecutorService get() {
      if (result == null) {
        result = Executors.newSingleThreadExecutor() 
      }
      return result;
    }
  };

**Comment by cgruber** [comment]
_From [gavin.k...@gmail.com](https://code.google.com/u/105743409453530897815/) on June 17, 2009 12:15:43_

"The idea of @Inject(optional=true) is that the bean will take advantage of a
dependency if there is one, otherwise it'll do its job using some meaningful default."

Not correct. Have you read the JavaDoc for @Inject. @Inject(optional=true) on a field
means "leave the field null". That's not a "meaningful" default.

"the bean would always call get(), then it'd have to handle getting a null back, or
worse an exception"

How is this different to having to handle a null value of the field, or a
NullPointerException?? 

In fact, having to call a method - get() - gives you a warning that you might need to
handle a null value/exception than having a field that might be null.

Urm, so from my point of view I could turn that round and say: "I wouldn't want all
beans that access an injected instance variable have to defend against nulls or
exceptions as a regular occurrence!"

**Comment by cgruber** [comment]
_From [gavin.k...@gmail.com](https://code.google.com/u/105743409453530897815/) on June 17, 2009 12:22:24_

"Yeah, Provider is not at all an alternative to optional=true. In fact, you can use
them together"

Are you (a) describing the documented semantics of Provider, (b) telling me that this
is what your product does today, or (c) making an argument that this is the right
thing to do?

If (a), I don't see where this is documented, and if it is, I'm not sure that it is
correct.

If (b), I don't care.

If (c), you have not actually made the argument.

Basically, the point I'm making, is that I would prefer that an "optional dependency"
be expressed by a non-null field of type Provider, than by a null field. The case of
"optional dependencies" is amazingly rare, IMO, and I therefore don't want to pollute
the @Inject annotation to handle this case.

**Comment by cgruber** [comment]
_From [gavin.k...@gmail.com](https://code.google.com/u/105743409453530897815/) on June 17, 2009 12:26:47_

"We could keep the feature and still drop optional from @Inject by adding a new type
for "faulty providers"."

This could be as simple as adding a binding type.

**Comment by cgruber** [comment]
_From [limpbizkit](https://code.google.com/u/limpbizkit/) on June 17, 2009 12:48:24_

gavin.king:
 \> Have you read the JavaDoc for @Inject. @Inject(optional=true) on a field means "leave the field null". That's 
not a "meaningful" default."

I don't know what spec you're reading, but the @Inject Javadoc says, "If a dependency matching the field can't be 
found, the injector will not set the field." It doesn't mention null at all. Therefore a default value (like 8080) may 
be used. http://atinject.googlecode.com/svn/trunk/javadoc/javax/inject/Inject.html

**Comment by cgruber** [comment]
_From [gavin.k...@gmail.com](https://code.google.com/u/105743409453530897815/) on June 17, 2009 13:15:49_

Ah, fair enough. It's true that optional=true is less objectionable *if* the field
has an explicit initializer. Unfortunately there's no way to limit its usage to
fields with initializers.

But seriously, is this thing *really* common enough that it really warrants
vandalizing the main @Inject annotation with a boolean annotation member? I *hate*
boolean-valued annotation members....

**Comment by cgruber** [comment]
_From [crazybob...@gmail.com](https://code.google.com/u/114815478402472606661/) on June 19, 2009 08:04:32_

Tim, could something similar to Guice's Modules.override() work as an alternative?

Optional injections have been one of the biggest sources of confusion for Guice users; it's difficult for them to 
differentiate between missing configuration vs. other types of errors. There's also the matter of whether we skip 
injections when transitive deps are missing vs. just the direct dep.

**Comment by cgruber** [comment]
_From [tpeie...@gmail.com](https://code.google.com/u/116658917155462694901/) on June 19, 2009 08:12:04_

[Bob Lee] Could something similar to Guice's Modules.override() work as an alternative?

I don't know enough about Modules.override() yet, but I bet you're right. I'll learn 
more and report whether my objection disappears.

**Comment by cgruber** [comment]
_From [tpeie...@gmail.com](https://code.google.com/u/116658917155462694901/) on June 22, 2009 08:11:19_

OK, I see how I could rewrite the DWR-Guice integration package (for example) to 
avoid @Inject(optional=true) using Modules.override:

    return Guice.createInjector(
        Modules.override(new DwrDefaultsModule()).with(/* user modules */));

It's not as convenient as being able to specify the default bindings on the fields 
themselves, but I do see that decoupling those defaults is a good thing in general.

But is specifying and mandating a facility like Modules.override really within the 
purview of JSR-330? If so, I cheerfully withdraw my objection. If not, well ... I 
wouldn't put up a fight about losing optional injection, but I wouldn't be cheerful 
about it, either.

**Comment by cgruber** [comment]
_From [crazybob...@gmail.com](https://code.google.com/u/114815478402472606661/) on June 22, 2009 11:56:09_

I'll definitely push for something like Modules.override() in the config API.

**Comment by cgruber** [comment]
_From [crazybob...@gmail.com](https://code.google.com/u/114815478402472606661/) on June 23, 2009 16:18:55_

It's gone as of r4 ( https://code.google.com/p/atinject/source/detail?r=4 ).

**Status:** Fixed

**Comment by cgruber** [comment]
_From [gavin.k...@gmail.com](https://code.google.com/u/105743409453530897815/) on June 23, 2009 18:36:15_

Great, thanks.

**Comment by cgruber** [comment]
_From [mccallum...@gmail.com](https://code.google.com/u/101543788066831713022/) on February 08, 2010 11:05:00_

Plese consider collection injection... see issue #20

**Comment by cgruber** [comment]
_From [ptit...@gmail.com](https://code.google.com/u/107779162194920573572/) on December 13, 2011 06:47:44_

Please reopen this, as there is no way to have collection injection!
